### PR TITLE
Addition of parameter definitions in MESSAGEix documentation

### LINKED ISSUE #260: documentation of parameters only lists dimensions, but lacks description
https://github.com/iiasa/message_ix/blob/71e52bd1b3635243adbca4a4b5ceddbc487807a8/message_ix/model/MESSAGE/parameter_def.gms#L18

With exception of very few 'general' parameters, a simple of description of parameters is not included in the MESSAGEix documentation and would be very helpful and straight forward to add as this information is included in the parameter definitions in the GAMS code.

### PR DESCRIPTION
**Pull request by GamzeUnlu95**
Closes #260.
Two other issues that are related and might also be addressed in this PR: #241 and #51.

This PR updates the parameter documentation and adds explanations for the parameters that are not self-explanatory. The parameter explanations are based on the definitions provided in `parameter_def.gms` file. 

Even though the initial idea was to add the definitions in the tables for every parameter, after a discussion with @volker-krey , we agreed that it is better to add these definitions only for parameters that are not clear since most of the parameters are already self-explanatory (e.g. bound_activity_up, construction_time etc.) and the tables look exhaustive (especially in the `pdf` version of the documentation) when all the definitions are added. 

The additional information and related links are added to the footnotes and to the explanation paragraphs for the parameters/sections that might require more explanation. (e.g. rating_bin, add-on technologies, cost parameters for ‘soft’ relaxations of dynamic constraints, auxiliary investment cost parameters and multipliers, emission scaling, relations, resources). 

These additions are based on the `parameter_def.gms` file which includes only basic definitions of the parameters. The reviewers could suggest changes to the explanations in case they are not clear enough (Especially for the section of **"Auxiliary investment cost parameters and multipliers"**).   

The documentation page can be built by using the command: `sphinx-build [source-directory] [output-directory]`

source-directory: `C:\....\message_ix\doc\source`  
output directory: `C:\....\message_ix\doc\build`

### CODE REVIEW

**Comment by khaeru** [comment]
@GamzeUnlu95 thanks for this. I have turned on the automatic RTD builds of this branch, and they seem to complete succesfully! Here is the page you modified: https://message.iiasa.ac.at/en/parameter_doc_new/model/MESSAGE/parameter_def.html

Please solicit others' input, in the form of comments on this PR, as to which (Option 1 or Option 2) is the clearest form for the documentation.

**Comment by khaeru** [comment]
Also note that, per our in-person discussion, message_ix tests are currently failing because iiasa/ixmp#225 needs to be completed, reviewed, and merged.

**Comment by khaeru** [comment]
@GamzeUnlu95 —commit 800c35a seems to revert the addition of descriptions and the Option 2 table that were added by bb8a35b. Was that intentional?

**Comment by GamzeUnlu95** [comment]
@khaeru , that was intentional. After converting to pdf and a discussion with Volker, we decided not to include all the definitions in the tables, but include them in the footnotes or in the paragraphs where it is relevant. I also updated the PR definition.

**Comment by francescolovat** [comment]
Hi @GamzeUnlu95,
thanks for you work. 

I've edited the description of the PR to include that closes issue #260. You've used the command _closes_ the other way around: you wrote closes #275 in the issue itself. The command closes issues from PRs. Therefore I also deleted that comment in #260.

Now I'm going through what you've wrote and I'll evidence very minor detail and also push a commit correcting a typo to also see if all the CIs pass after iiasa/ixmp#225 was merged.

File: message_ix/model/MESSAGE/parameter_def.gms [hunk 1]
```diff
@@ -200,9 +206,14 @@ Parameter
...
-*    (:ref:`reliability_constraint`) and flexibility (:ref:`flexibility_constraint`) can depend on the share
-*    of the technology output in the total commodity use at a specific level.
+* .. [#rating] Maximum share of technology in commodity use per rating. The upper bound of a contribution by any technology to the constraints on system reliability
+*    (:ref:`reliability_constraint`) and flexibility (:ref:`flexibility_constraint`) can depend on the share of the technology output in the total commodity use at
+*    a specific level.
+*
+* .. [#flexfactor] Contribution of technologies towards operation flexibility constraint. It is used in :ref:`flexibility_constraint`.
+*
+* .. [#renewables] ``renewable_capacity_factor`` refers to the quality of renewable potential by grade and ``renewable_potential`` refers to the size of the renewable potential per grade. 
+*
```
Comment (francescolovat) [inline message_ix/model/MESSAGE/parameter_def.gms]: I don't understand the sentence: _refers to the quality of renewable potential by grade_ here. Do you think it is enough? Probably I missed other previous descriptions that could help me understand it.

File: message_ix/model/MESSAGE/parameter_def.gms [hunk 2]
```diff
@@ -73,6 +73,11 @@ Parameters
 * Parameters of the `Resources` section
 * -------------------------------------
 *
+* The quantity of the resources in |MESSAGEix| is represented by two parameters: ``resource_volume`` and ``resource_remaining``.
+* The first parameter is the volume of resources at the start of the model horizon while the second is the maximum extraction relative
+* to remaining resources by year. The maximum extraction of the resources (by grade) in a year is constrained by the parameter
```
Comment (francescolovat) [inline message_ix/model/MESSAGE/parameter_def.gms]: I don't understand the sentence: _while the second is the maximum extraction relative to remaining resources by year_ here. Do you think it is enough? Probably I missed other previous descriptions that could help me understand it.

Shouldn't it be that the second are just the remaining sources per year?

File: message_ix/model/MESSAGE/parameter_def.gms [hunk 3]
```diff
@@ -450,17 +452,17 @@ Parameters
...
 * ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 *
-* Auxilary investment cost parameters include the remaining technical lifetime at the end of model horizon in addition to the
-* different scaling factors and  multipliers as listed below. These factors account for remaining capacity or construction time of new capcaity, 
-* the value of investment at the end of mdoel horizon or the discount factor of remaining lifetime beyond model horizon. 
-* 
+* Auxiliary investment cost parameters include the remaining technical lifetime at the end of model horizon (``beyond_horizon_lifetime``) in addition to the
+* different scaling factors and multipliers as listed below. These factors account for remaining capacity (``remaining_capacity``) or construction time of new capacity (``construction_time_factor``),
+* the value of investment at the end of model horizon (``end_of_horizon_factor``) or the discount factor of remaining lifetime beyond model horizon (``beyond_horizon_factor``).
+*
```
Comment (francescolovat) [inline message_ix/model/MESSAGE/parameter_def.gms]: please @GamzeUnlu95 have a look if I inserted the right parameters into the right pieces of the paragraph. It was not clear for me that's why I wanted to explicitly relate the text to the parameters listed below.

**Comment by francescolovat** [comment]
Looking for some other wrong parameter definitions in [Parameter definition page](https://message.iiasa.ac.at/en/master/model/MESSAGE/parameter_def.html), as emphasized in #51, I have a doubt now:

- which is the difference between `node_loc` and `node` used in the tables in the [Parameter definition page](https://message.iiasa.ac.at/en/master/model/MESSAGE/parameter_def.html)? `node_loc` is not even defined in the [Sets definition page](https://message.iiasa.ac.at/en/master/model/MESSAGE/sets_maps_def.html), so I don't really know why it's being used to describe parameters and what differentiates it from `node`. 

I thought probably the aim was to differentiate the _sets_ **location** and **node** from the `model_core.gms` file. But, as you can see in the snapshot below, both of them have indistinctly been used to describe parameters having **node** and **location** as dimensions in `model_core.gms`.

Can anyone help me interpret them so I can properly edit their dimensions and eventually close #51 ? @OFR-IIASA @volker-krey @khaeru  

image

**Comment by khaeru** [comment]
@francescolovat —that's a great question. Let's also CC @danielhuppmann @gidden per #254.

There are at least three things going on here:

1. In ixmp (Python/Java code), every dimension of a parameter has both an index **set** and an optional **index/dimension name**.
   - See e.g. the [init_par() docs](https://message.iiasa.ac.at/projects/ixmp/en/master/api-python.html#ixmp.Scenario.init_par).
   - The *sets* must be actual sets in the Scenario.
   - The *names* can be anything; e.g. you can have a dimension indexed by `year` but named `'foo'`.
2. In the MESSAGE GAMS code, there are a number of **aliases for sets**.
   - See [sets_maps_def.gms](https://github.com/iiasa/message_ix/blob/master/message_ix/model/MESSAGE/sets_maps_def.gms).
   - For instance, the lines: https://github.com/iiasa/message_ix/blob/d7a75e63e7d16632c7bd0ed88b0a220b70a8c42b/message_ix/model/MESSAGE/sets_maps_def.gms#L133-L136 show that e.g. `node_share` is always an alias of `node`.
   - This is needed because a GAMS parameter that has the *same* index set for *two* of its dimensions can't re-use the same name, e.g. you can't define a GAMS `parameter foo(year, year);`.
3. See #254. While the GAMS definition of MESSAGE's structure is in the .gms files, there is (must be) a parallel definition of the same structure for the Python code to use. At the moment, this is (wrongly, IMO) hidden in the non-open ixmp_source Java package.
   - See for instance here: https://github.com/iiasa/ixmp_source/blob/6c2afd932e10e284ff12c4a20a0d62372f609228/src/main/java/at/ac/iiasa/ixmp/modelspecs/MESSAGEspecs.java#L877 and here: https://github.com/iiasa/ixmp_source/blob/6c2afd932e10e284ff12c4a20a0d62372f609228/src/main/java/at/ac/iiasa/ixmp/modelspecs/MESSAGEspecs.java#L76-L82
    - These lines are basically equivalent to: `scen.init_par('historical_new_capacity', idx_sets=['node', 'technology', 'year'], idx_names=['node_loc', 'technology', 'year_vtg'])`

So what you've uncovered (I hadn't noticed myself!) is:
- There is not necessarily a correspondence between a **dimension name** on the Python side and any **alias** on the GAMS side.
  - E.g. `node_loc` is a dimension name, but there is no GAMS alias.
  - E.g. `land_emission` is defined over the GAMS alias `location`, but on the Python side it does not have a dimension name.
- Some index names, like `node_loc`, are **hidden in the Java code** and thus the correspondence between them and particular sets can only be guessed at.

I hope this helps to understand the situation and to propose improvements.

**Comment by francescolovat** [comment]
Yes, thanks a lot @khaeru , I didn't know about that aliases in GAMS were needed. I'll go through the links you suggested to understand the whole think better. Btw, I've just realized that the last commit trying to help solving the issue #51 has no sense and it's wrong. I'll revert it now.

**Comment by danielhuppmann** [comment]
To quote Douglas Adams: "I apologize for the inconvenience".

There are indeed many things going on here, and it is important to keep two things separate:

1) **data handling** - think of a `pandas.DataFrame` representing parameter data
   Here we need to have:
   - **consistency checking**, so that an entry in the node-column (e.g., where a technology is located) exists in the set `node`. This is implemented via `index_set` (and can be overridden by setting `index_set='*'`).
   - **column naming**: each column of a dataframe has to have a unique name so that we easily work with it. The column names should be a trade-off between intelligibility and consistency (for example using `node_loc` for the column that specifies a location for all parameters related to a technology).
2) **mathematical formulation**
    The equations of the optimisation problem are quite elaborate to make life easier for a user. For example, the emissions bound of a node automatically considers the emissions of all sub-nodes - otherwise, a user would have to specify specific bounds of sub-nodes (determining the allocation of the constraint between sub-nodes a priori), or make some other elaborate parameter specifications.
    This is implemented in GAMS using multiple nested loops, and this requires aliases to correctly implement the summations or if-else-statements. The primary goal of the alias naming is to make it easy for users to understand an equation.

These two use cases are, in principle, not related.

**Pull request merged**